Added cuBLAS path for torch.linalg.lstsq

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54725 Added cuBLAS path for torch.linalg.lstsq**
* #57317 Add cuSOLVER path for torch.linalg.lstsq

cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Performance comparison cuSOLVER vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-832234456.
Performance comparison MAGMA vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039.

Differential Revision: [D28248803](https://our.internmc.facebook.com/intern/diff/D28248803)

IvanYashchuk committed:
Added cuBLAS path for torch.linalg.lstsq

cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Ref. https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry there was an illegal memory error for CUDA 10 (version 11 worked locally) causing CI failures. We still need to wait for CI builds to finish but it should be fixed now.

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

mruberry commented:
@xwang233 would you take a look at this cuBLAS update? 

@IvanYashchuk the rest of the stack looks ready to merge but may need a rebase. Just ping me when I should start merging it.

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I rebased this stack. Hopefully, there won't be conflicts with other stacks of PRs that are getting merged.

xwang233 commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -106,6 +106,74 @@ void triangular_solve_batched_cublas(Tensor& A, Tensor& B, Tensor& infos, bool u
+    m >= n,
+    "torch.linalg.lstsq: only overdetermined systems (input.size(-2) >= input.size(-1)) are allowed on CUDA with cuBLAS backend.");
+
+  // cuBLAS documentation says:
+  // Matrices Aarray[i] should not overlap; otherwise, undefined behavior is expected.
+  // explicitly broadcast the batch dimensions of A
+  IntArrayRef A_batch_sizes(A.sizes().data(), A.dim() - 2);
+  IntArrayRef B_batch_sizes(B.sizes().data(), B.dim() - 2);
+  std::vector<int64_t> expand_batch_portion = at::infer_size(A_batch_sizes, B_batch_sizes);
+  expand_batch_portion.insert(expand_batch_portion.end(), {A.size(-2), A.size(-1)});
+  Tensor A_expanded = A.expand({expand_batch_portion});
+  Tensor A_broadcasted = cloneBatchedColumnMajor(A_expanded);
```
I'm a bit concerned with this broadcasting behavior. (which is a really nice feature!) Do we have a documentation entry mentioning about the linear algebra broadcasting (for batch dimension)? Also, can you please check if we have a test for this broadcasting?

I think linalg broadcasting sounds like a big feature that we should either enable for most operators, or just don't mention in the doc. 😅

IvanYashchuk replied:
There is a test `test_linalg_lstsq_batch_broadcasting` in `test/test_linalg.py`
https://github.com/pytorch/pytorch/blob/f84f2063b43a32fbe88c7e4de7e1019a00adb302/test/test_linalg.py#L250

Broadcasting is applicable only for solve functions because they are the only ones that take two interacting inputs.
I don't think that it is documented separately somewhere, it's just expected that this function is broadcastable as many others in PyTorch are. lstsq still doesn't implement broadcasting if some dimensions do not exist (general behavior described here https://pytorch.org/docs/stable/notes/broadcasting.html). Other solve functions materialize explicitly the broadcasted inputs as done here for cuBLAS, and lstsq is the first function that tries to address the memory inefficiencies.

We have two issues open for discussing the broadcasting of linear algebra solvers https://github.com/pytorch/pytorch/issues/52915 and https://github.com/pytorch/pytorch/issues/49252.

xwang233 replied:
Thanks. That's nice!

xwang233 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -2695,14 +2695,32 @@ static void apply_gels(const Tensor& a, Tensor& b, Tensor& infos) {
 
+void gels_magma(const Tensor& a, Tensor& b, Tensor& infos) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(a.scalar_type(), "gels_magma", [&] {
+    apply_gels<scalar_t>(a, b, infos);
```
MAGMA code is not shown in this PR. Does that support broadcasting as well?

IvanYashchuk replied:
It does, here are the tests https://github.com/pytorch/pytorch/blob/f84f2063b43a32fbe88c7e4de7e1019a00adb302/test/test_linalg.py#L250

It uses @nikitaved's `batch_iterator_with_broadcasting` and it's easy there as MAGMA calls are device synchronizing.

xwang233 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -2695,14 +2695,32 @@ static void apply_gels(const Tensor& a, Tensor& b, Tensor& infos) {
+#else
+    gels_magma(a, b, infos);
+#endif // USE_MAGMA
+  }
```
Do you have a benchmark table for this?

IvanYashchuk replied:
Posted it here https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039.

xwang233 reviewed (commented):
Thanks for the work @IvanYashchuk ! This PR overall looks good. I've left some comments about broadcasting.

IvanYashchuk commented:
Here are the performance results for float32 inputs, the results are only for the square matrices and the right-hand-size tensor has one column.
```
|                              | cuBLAS | MAGMA |
|------------------------------|--------|-------|
| torch.Size([32, 32])         | 0.6    | 6     |
| torch.Size([1, 32, 32])      | 0.5    | 6     |
| torch.Size([32, 32, 32])     | 0.5    | 180   |
| torch.Size([64, 32, 32])     | 0.5    | 400   |
| torch.Size([64, 64])         | 1.5    | 6     |
| torch.Size([1, 64, 64])      | 1.5    | 6     |
| torch.Size([32, 64, 64])     | 1.8    | 200   |
| torch.Size([64, 64, 64])     | 1.8    | 400   |
| torch.Size([128, 128])       | 6.4    | 7     |
| torch.Size([1, 128, 128])    | 6.4    | 7     |
| torch.Size([32, 128, 128])   | 8.1    | 210   |
| torch.Size([64, 128, 128])   | 8.2    | 400   |
| torch.Size([256, 256])       | 29     | 8     |
| torch.Size([1, 256, 256])    | 29     | 8     |
| torch.Size([32, 256, 256])   | 43.9   | 200   |
| torch.Size([64, 256, 256])   | 48.7   | 500   |
| torch.Size([512, 512])       | 144    | 10    |
| torch.Size([1, 512, 512])    | 143.8  | 10    |
| torch.Size([32, 512, 512])   | 289.1  | 300   |
| torch.Size([64, 512, 512])   | 303    | 600   |
| torch.Size([1024, 1024])     | 807.6  | 16    |
| torch.Size([1, 1024, 1024])  | 807.6  | 20    |
| torch.Size([32, 1024, 1024]) | 2715   | 500   |
| torch.Size([64, 1024, 1024]) | 2800   | 1000  |

Times are in milliseconds (ms).
```

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk commented:
Note: this PR needs to be rebased and combined with another stack that implements cuSOLVER path for torch.linalg.lstsq https://github.com/pytorch/pytorch/pull/57317

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.


[ghstack-poisoned]

IvanYashchuk commented:
Comparing cuBLAS vs cuSOLVER for float32 inputs, the results are only for the square matrices and the right-hand-size tensor has one column.

<details>
  <summary>Comparison table, batches 1-64 and sizes 32x32 - 1024x1024</summary>
  
 ```
Times are in microseconds (us).

|                              | cuSOLVER | cuBLAS  |
|------------------------------|----------|---------|
| torch.Size([32, 32])         | 520      | 860     |
| torch.Size([1, 32, 32])      | 500      | 690     |
| torch.Size([32, 32, 32])     | 2000     | 733     |
| torch.Size([64, 32, 32])     | 3100     | 733     |
| torch.Size([64, 64])         | 900      | 1800    |
| torch.Size([1, 64, 64])      | 900      | 1800    |
| torch.Size([32, 64, 64])     | 15000    | 2090    |
| torch.Size([64, 64, 64])     | 24000    | 2090    |
| torch.Size([128, 128])       | 2000     | 6800    |
| torch.Size([1, 128, 128])    | 1500     | 6810    |
| torch.Size([32, 128, 128])   | 30000    | 8391    |
| torch.Size([64, 128, 128])   | 57000    | 9430    |
| torch.Size([256, 256])       | 2000     | 29300   |
| torch.Size([1, 256, 256])    | 1000     | 29310   |
| torch.Size([32, 256, 256])   | 59400    | 48130   |
| torch.Size([64, 256, 256])   | 72900    | 51000   |
| torch.Size([512, 512])       | 2870     | 145600  |
| torch.Size([1, 512, 512])    | 2900     | 146000  |
| torch.Size([32, 512, 512])   | 79300    | 295000  |
| torch.Size([64, 512, 512])   | 157000   | 316000  |
| torch.Size([1024, 1024])     | 6220     | 972300  |
| torch.Size([1, 1024, 1024])  | 6300     | 973700  |
| torch.Size([32, 1024, 1024]) | 190000   | 2700000 |
| torch.Size([64, 1024, 1024]) | 377500   | 2880000 |
  ```
  
</details>

The following results show that for inputs smaller than 256x256 cuBLAS is faster than cuSOLVER when `batch_count(input) >= max(2, input.shape[-2] / 16)`

<details>
  <summary>Comparison table, batches 1-33 and sizes 32x32 - 256x256</summary>
  
```
|                            | cuSOLVER | cuBLAS |
|----------------------------|----------|--------|
| torch.Size([1, 32, 32])    | 490      | 860    |
| torch.Size([2, 32, 32])    | 767      | 692    | <- batchsize >= 32 / 16
| torch.Size([3, 32, 32])    | 810      | 700    |
| torch.Size([4, 32, 32])    | 850      | 702    |
| torch.Size([5, 32, 32])    | 890      | 709    |
| torch.Size([6, 32, 32])    | 930      | 710    |
| torch.Size([7, 32, 32])    | 973      | 715    |
| torch.Size([8, 32, 32])    | 1020     | 726    |
| torch.Size([9, 32, 32])    | 1050     | 723    |
| torch.Size([10, 32, 32])   | 1090     | 723    |
| torch.Size([11, 32, 32])   | 1100     | 724    |
| torch.Size([12, 32, 32])   | 1160     | 720    |
| torch.Size([13, 32, 32])   | 1190     | 722    |
| torch.Size([14, 32, 32])   | 1200     | 721    |
| torch.Size([15, 32, 32])   | 1300     | 720    |
| torch.Size([16, 32, 32])   | 1300     | 724    |
| torch.Size([17, 32, 32])   | 1300     | 724    |
| torch.Size([18, 32, 32])   | 1370     | 730    |
| torch.Size([19, 32, 32])   | 1410     | 725    |
| torch.Size([20, 32, 32])   | 1438     | 725    |
| torch.Size([21, 32, 32])   | 1500     | 720    |
| torch.Size([22, 32, 32])   | 1500     | 722    |
| torch.Size([23, 32, 32])   | 1540     | 730    |
| torch.Size([24, 32, 32])   | 1600     | 724    |
| torch.Size([25, 32, 32])   | 1610     | 720    |
| torch.Size([26, 32, 32])   | 1600     | 723    |
| torch.Size([27, 32, 32])   | 1680     | 725    |
| torch.Size([28, 32, 32])   | 1710     | 724    |
| torch.Size([29, 32, 32])   | 1800     | 724    |
| torch.Size([30, 32, 32])   | 1800     | 724    |
| torch.Size([31, 32, 32])   | 1800     | 730    |
| torch.Size([32, 32, 32])   | 1850     | 723    |
|----------------------------|----------|--------|
| torch.Size([1, 64, 64])    | 800      | 1803   |
| torch.Size([2, 64, 64])    | 1310     | 1824   |
| torch.Size([3, 64, 64])    | 1800     | 1850   |
| torch.Size([4, 64, 64])    | 2930     | 1917   |  <- batchsize >= 64 / 16
| torch.Size([5, 64, 64])    | 3300     | 1990   |
| torch.Size([6, 64, 64])    | 3700     | 2042   |
| torch.Size([7, 64, 64])    | 4050     | 2051   |
| torch.Size([8, 64, 64])    | 4400     | 2080   |
| torch.Size([9, 64, 64])    | 4700     | 2090   |
| torch.Size([10, 64, 64])   | 5000     | 2076   |
| torch.Size([11, 64, 64])   | 5310     | 2080   |
| torch.Size([12, 64, 64])   | 5620     | 2083   |
| torch.Size([13, 64, 64])   | 5900     | 2100   |
| torch.Size([14, 64, 64])   | 6230     | 2085   |
| torch.Size([15, 64, 64])   | 6550     | 2084   |
| torch.Size([16, 64, 64])   | 6900     | 2090   |
| torch.Size([17, 64, 64])   | 7100     | 2088   |
| torch.Size([18, 64, 64])   | 7500     | 2090   |
| torch.Size([19, 64, 64])   | 7800     | 2090   |
| torch.Size([20, 64, 64])   | 8100     | 2087   |
| torch.Size([21, 64, 64])   | 8380     | 2086   |
| torch.Size([22, 64, 64])   | 8680     | 2080   |
| torch.Size([23, 64, 64])   | 9000     | 2080   |
| torch.Size([24, 64, 64])   | 9300     | 2100   |
| torch.Size([25, 64, 64])   | 9640     | 2080   |
| torch.Size([26, 64, 64])   | 9900     | 2090   |
| torch.Size([27, 64, 64])   | 10300    | 2080   |
| torch.Size([28, 64, 64])   | 10500    | 2090   |
| torch.Size([29, 64, 64])   | 11000    | 2100   |
| torch.Size([30, 64, 64])   | 11210    | 2080   |
| torch.Size([31, 64, 64])   | 11500    | 2090   |
| torch.Size([32, 64, 64])   | 11800    | 2100   |
|----------------------------|----------|--------|
| torch.Size([1, 128, 128])  | 1400     | 6850   |
| torch.Size([2, 128, 128])  | 2500     | 7660   |
| torch.Size([3, 128, 128])  | 3600     | 7811   |
| torch.Size([4, 128, 128])  | 4630     | 7893   |
| torch.Size([5, 128, 128])  | 5680     | 8150   |
| torch.Size([6, 128, 128])  | 6800     | 8181   |
| torch.Size([7, 128, 128])  | 7800     | 8292   |
| torch.Size([8, 128, 128])  | 13100    | 8417   |  <- batchsize >= 128 / 16
| torch.Size([9, 128, 128])  | 14000    | 8430   |
| torch.Size([10, 128, 128]) | 14000    | 8420   |
| torch.Size([11, 128, 128]) | 14900    | 8415   |
| torch.Size([12, 128, 128]) | 15500    | 8418   |
| torch.Size([13, 128, 128]) | 16100    | 8440   |
| torch.Size([14, 128, 128]) | 16600    | 8420   |
| torch.Size([15, 128, 128]) | 17000    | 8442   |
| torch.Size([16, 128, 128]) | 17800    | 8430   |
| torch.Size([17, 128, 128]) | 18500    | 8420   |
| torch.Size([18, 128, 128]) | 19000    | 8420   |
| torch.Size([19, 128, 128]) | 20000    | 8435   |
| torch.Size([20, 128, 128]) | 20200    | 8430   |
| torch.Size([21, 128, 128]) | 20900    | 8415   |
| torch.Size([22, 128, 128]) | 21000    | 8440   |
| torch.Size([23, 128, 128]) | 22000    | 8400   |
| torch.Size([24, 128, 128]) | 23000    | 8450   |
| torch.Size([25, 128, 128]) | 23200    | 8422   |
| torch.Size([26, 128, 128]) | 24000    | 8420   |
| torch.Size([27, 128, 128]) | 24400    | 8473   |
| torch.Size([28, 128, 128]) | 25000    | 8432   |
| torch.Size([29, 128, 128]) | 25600    | 8473   |
| torch.Size([30, 128, 128]) | 26000    | 8450   |
| torch.Size([31, 128, 128]) | 26800    | 8493   |
| torch.Size([32, 128, 128]) | 27000    | 8474   |
|----------------------------|----------|--------|
| torch.Size([1, 256, 256])  | 1390     | 29520  |
| torch.Size([2, 256, 256])  | 2500     | 35590  |
| torch.Size([3, 256, 256])  | 3570     | 35890  |
| torch.Size([4, 256, 256])  | 4655     | 36230  |
| torch.Size([5, 256, 256])  | 5750     | 36820  |
| torch.Size([6, 256, 256])  | 6840     | 37230  |
| torch.Size([7, 256, 256])  | 7930     | 37870  |
| torch.Size([8, 256, 256])  | 9010     | 39030  |
| torch.Size([9, 256, 256])  | 10150    | 38830  |
| torch.Size([10, 256, 256]) | 11200    | 38860  |
| torch.Size([11, 256, 256]) | 12300    | 39060  |
| torch.Size([12, 256, 256]) | 13380    | 39210  |
| torch.Size([13, 256, 256]) | 14500    | 41330  |
| torch.Size([14, 256, 256]) | 15540    | 42900  |
| torch.Size([15, 256, 256]) | 16620    | 43420  |
| torch.Size([16, 256, 256]) | 49000    | 44010  |  <- batchsize >= 256 / 16
| torch.Size([17, 256, 256]) | 50330    | 44580  |
| torch.Size([18, 256, 256]) | 50800    | 44810  |
| torch.Size([19, 256, 256]) | 51200    | 45000  |
| torch.Size([20, 256, 256]) | 51700    | 45600  |
| torch.Size([21, 256, 256]) | 52600    | 45890  |
| torch.Size([22, 256, 256]) | 53360    | 46110  |
| torch.Size([23, 256, 256]) | 53800    | 46360  |
| torch.Size([24, 256, 256]) | 54470    | 46710  |
| torch.Size([25, 256, 256]) | 55200    | 47100  |
| torch.Size([26, 256, 256]) | 55350    | 47180  |
| torch.Size([27, 256, 256]) | 56470    | 47110  |
| torch.Size([28, 256, 256]) | 56400    | 47640  |
| torch.Size([29, 256, 256]) | 57400    | 47600  |
| torch.Size([30, 256, 256]) | 57660    | 48100  |
| torch.Size([31, 256, 256]) | 58350    | 48060  |
| torch.Size([32, 256, 256]) | 58900    | 48500  |
```
</details>

<details>
  <summary>Code to produce one column of results (I recompiled PyTorch for each column) </summary>
  
 ```python
import numpy, torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys

print('Using pytorch %s' % (torch.__version__))
print('Using numpy %s' % (numpy.__version__))

numpy.random.seed(1234)

shapes = [32, 64, 128, 256, 512, 1024]#, 2048]#, 4096, 8192]
batches = [(), (1,), (32,), (64,)]
results = []
num_threads = 1
dtype = torch.float32
repeats = 2

from torch.testing._internal.common_utils import random_hermitian_pd_matrix


for shape, batch in itertools.product(shapes, batches):
    a = torch.randn(*batch, shape, shape, dtype=dtype, device='cuda')
    b = torch.randn(*batch, shape, 1, dtype=dtype, device='cuda')

    a_cpu = a.to('cpu')
    b_cpu = b.to('cpu')

    # tasks = [("numpy.linalg.lstsq(a_cpu, b_cpu, rcond=-1)", "lstsq NumPy"), ("torch.linalg.lstsq(a_cpu, b_cpu)", "lstsq CPU"), ("torch.linalg.lstsq(a, b)", "lstsq CUDA"), ]
    tasks = [("torch.linalg.lstsq(a, b)", "lstsq cuSOLVER"),]
    timers = [Timer(stmt=stmt, num_threads=num_threads, label=f"lstsq (ATen) {dtype}", sub_label=f"{a.shape}", description=label, globals=globals()) for stmt, label in tasks]

    for i, timer in enumerate(timers * repeats):
        results.append(
            timer.blocked_autorange()
        )
        print(f"\r{i + 1} / {len(timers) * repeats}", end="")
        sys.stdout.flush()

compare = Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()

  ```
  
</details>

cuBLAS vs MAGMA: https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Performance comparison cuSOLVER vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-832234456.
Performance comparison MAGMA vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Performance comparison cuSOLVER vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-832234456.
Performance comparison MAGMA vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039.

[ghstack-poisoned]

mruberry commented:
Rest of this stack looks good; how's this diff going, @IvanYashchuk, @xwang233?

I'll start landing the other PRs in the stack now.

xwang233 reviewed (approved):
LGTM, thanks for the work!

mruberry reviewed (approved):
Stamped; thanks @IvanYashchuk, @xwang233!

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248803).

IvanYashchuk committed:
Update on "Added cuBLAS path for torch.linalg.lstsq"


cuBLAS's gelsBatched is faster than MAGMA's for matrices with rows less
than 128.

Performance comparison cuSOLVER vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-832234456.
Performance comparison MAGMA vs cuBLAS: https://github.com/pytorch/pytorch/pull/54725#issuecomment-827649039.

Differential Revision: [D28248803](https://our.internmc.facebook.com/intern/diff/D28248803)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248803).
